src/hotspot/share/opto/escape.cpp

```diff
@@ -502,9 +502,10 @@ bool ConnectionGraph::can_reduce_phi_check_users(PhiNode* ophi) const {
 // details.
 bool ConnectionGraph::can_reduce_phi(PhiNode* ophi) const {
   // If there was an error attempting to reduce allocation merges for this
-  // method we might have disabled the compilation and be retrying
-  // with RAM disabled.
-  if (!_compile->do_reduce_allocation_merges()) {
+  // method we might have disabled the compilation and be retrying with RAM
+  // disabled.
+  // If EliminateAllocations is False, there is no point in reducing merges.
+  if (!ReduceAllocationMerges || !EliminateAllocations || !_compile->do_reduce_allocation_merges()) {
```

Comment:
1. You don't need to check `ReduceAllocationMerges` because `do_reduce_allocation_merges` is set by its value.
2. Setting `do_reduce_allocation_merges` in `C2Compiler::compile_method()` should also check `EliminateAllocations` flag. Then you don't need to check this flag here.